reflex/components/core/match.py

```diff
@@ -0,0 +1,243 @@
+
+        Raises:
+            ValueError: If there are multiple default cases.
+        """
+        default = None
+
+        if len([case for case in cases if not isinstance(case, tuple)]) > 1:
+            raise ValueError("rx.match can only have one default case.")
+
+        # Get the default case which should be the last non-tuple arg
+        if not isinstance(cases[-1], tuple):
+            default = cases.pop()
+            default = (
+                Var.create(default)
+                if not isinstance(default, BaseComponent)
+                else default
+            )
+
+        return cases, default  # type: ignore
+
+    @classmethod
+    def _process_match_cases(cls, cases: List) -> List[List[BaseVar]]:
+        """Process the individual match cases.
+
+        Args:
+            cases: The match cases.
+
+        Returns:
+            The processed match cases.
+
+        Raises:
+            ValueError: If the default case is not the last case or the tuple elements are less than 2.
+        """
+        match_cases = []
+        for case in cases:
+            if not isinstance(case, tuple):
+                raise ValueError(
+                    "rx.match should have tuples of cases and a default case as the last argument."
+                )
+            # There should be at least two elements in a case tuple(a condition and return value)
+            if len(case) < 2:
+                raise ValueError(
+                    "A case tuple should have at least a match case element and a return value."
+                )
+
+            case_list = []
+            for element in case:
+                # convert all non component element to vars.
+                el = (
+                    Var.create(element)
+                    if not isinstance(element, BaseComponent)
+                    else element
+                )
+                if not isinstance(el, (BaseVar, BaseComponent)):
+                    raise ValueError("Case element must be a var or component")
+                case_list.append(el)
+
+            match_cases.append(case_list)
+
+        return match_cases
+
+    @classmethod
+    def _validate_return_types(cls, match_cases: List[List[BaseVar]]) -> None:
+        """Validate that match cases have the same return types.
+
+        Args:
+            match_cases: The match cases.
+
+        Raises:
+            TypeError: If the return types of cases are different.
+        """
+        first_case_return = match_cases[0][-1]
+        return_type = type(first_case_return)
+
+        if types._isinstance(first_case_return, BaseComponent):
+            return_type = BaseComponent
+        elif types._isinstance(first_case_return, BaseVar):
+            return_type = BaseVar
+
+        for case in match_cases:
+            if not types._issubclass(type(case[-1]), return_type):
+                raise TypeError("match cases should have the same return types")
```

Comment:
Include the checked types here, otherwise it might be hard to track down the exact issue.

```python
raise TypeError(f"Match cases should have the same return types. Case {index} of type {type(case[-1])!r} is not {return_type}")
```